### ISSUE TITLE
S3 request URL is constructed inconsistently

### ISSUE THREAD

**Issue by Fang-** [urbit/urbit#3463]
Set up a local S3 server using [minio](https://min.io/). Took some fiddling to get this to work. I observed something weird.

With an endpoint of `http://plfn.io:9000` and a bucket named `img`, it tries to make requests to `http://img.plfn.io:9000/palfun-foslup/filename.png`.

With an endpoint of `http://plfn.io:9000` and a bucket named `h`, it tries to make requests to `http://plfn.io:9000/h/palfun-foslup/filename.png`.

It seems that some bucket names get put as the subdomain, while others get put as part of the path. From my now-working configuration, it seems the latter is correct behavior, making the second case correct, and the first case use a bucket named `palfun-foslup` instead.

(I want uploads to go to the root of the bucket, instead of a shipname folder anyway, but I guess that's outside the scope of this issue.)

**Comment by mike-cr** [comment]
AWS are deprecating path-style requests at the end of this month afaik, so the subdomain style should be the correct one. You can enable this in minio with the `MINIO_DOMAIN="plfn.io"` env-var in `/etc/minio/minio.conf`

**Comment by matildepark** [comment]
> (I want uploads to go to the root of the bucket, instead of a shipname folder anyway, but I guess that's outside the scope of this issue.)

This was because we could easily imagine a world where your s3 bucket is configured by your hosting provider or star and thus we want to scope the bucket.

**Comment by matildepark** [comment]
@Fang- — alleviated after @mike-cr 's instructions?

**Comment by Fang-** [comment]
My setup is terrible and doesn't let me edit the config (d-don't ask), but the instructions clear up my confusion around expected behavior here, yes.

The inconsistent construction of the URL (ie path vs subdomain) based on (seemingly) bucket name length is probably not 100% correct though?

**Comment by tylershuster** [comment]
This is not a landscape issue — all landscape does is say "hey bucket (marpem-images), here's a file (blob), put it in this path (radbur-sivmus/file.png)" and then the api sends back the URL of the uploaded file.

**Comment by matildepark** [comment]
I agree that I think this is a config issue, but I do want @Fang- to have a nice time.

**Comment by Fang-** [comment]
I'm not talking about the config/setup side of the issue anymore though. (^:

> all landscape does is say "hey bucket (marpem-images), here's a file (blob), put it in this path (radbur-sivmus/file.png)" and then the api sends back the URL of the uploaded file.

Sure. What I'm saying is that the way landscape calls out to a specific bucket seems to be inconsistent. The place where it puts the bucket name in the request depends on (seemingly) the length of the bucket name:

> With an endpoint of http://plfn.io:9000 and a bucket named **`img`**, it tries to make requests to `http://`**`img.`**`plfn.io:9000/palfun-foslup/filename.png`.
> 
> With an endpoint of http://plfn.io:9000 and a bucket named **`h`**, it tries to make requests to `http://plfn.io:9000/`**`h/`**`palfun-foslup/filename.png`.

Notice how in one case, the bucket name goes in the subdomain. In the other case, the bucket name goes in the url path.

**Comment by mike-cr** [comment]
The AWS S3 API expects bucket names to be [between 3 and 63 characters long](https://docs.aws.amazon.com/AmazonS3/latest/dev/BucketRestrictions.html)

I would guess that the path-style url for shorter bucket names is a result of the aws-sdk trying to handle the invalid bucket

**Comment by mike-cr** [comment]
you could probably either validate the bucket name input, or maybe you could include the bucket name in the `endpoint` in `setCredentials` and add `bucket_endpoint: true`. That would theoretically force subdomain style, dunno if it would break anything
